support for electron 32

With electron 32.0.1 and its underlying v8 version 12.8, there are mulitple errors when building NAN modules:

nan.h(700,39): error C2039: "IdleNotificationDeadline" ist kein Member von "v8::Isolate"
nan.h(2560,8): error C2039: "SetAccessor" ist kein Member von "v8::ObjectTemplate"
nan.h(2654,15): error C2039: "SetAccessor" ist kein Member von "v8::Object"
nan.h(2730,56): error C2440: "<function-style-cast>": "initializer list" kann nicht in "v8::NamedPropertyHandlerConfiguration" konvertiert werden
nan_scriptorigin.h(19,1): error C2664: "v8::ScriptOrigin::ScriptOrigin(v8::Local<v8::Value>,int,int,bool,int,v8::Local<v8::Value>,bool,bool,bool,v8::Local<v8::Data>)" : Konvertierung von Argument 1 von "v8::Isolate *" in "v8::Local<v8::Value>" nicht möglich

In Electron v31 the following lines appeared as warnings

```
node_modules\nan\nan.h(700,39): warning C4996: 'v8::Isolate::IdleNotificationDeadline' : Use MemoryPressureNotification() to influence the GC schedule. 
node_modules\nan\nan.h(2654,15): warning C4996: 'v8::Object::SetAccessor': Use SetNativeDataProperty instead 
node_modules\nan\nan.h(2730,56): warning C4996: 'v8::NamedPropertyHandlerConfiguration ::NamedPropertyHandlerConfiguration': Provide interceptor callbacks with new signatures instead (NamedPropertyXxxCallback)
node_modules\nan\nan_scriptorigin.h(19,23): warning C4996: 'v8::ScriptOrigin::ScriptOrigin': Use constructor without the isolate. 
node_modules\nan\nan_scriptorigin.h(23,23): warning C4996: 'v8::ScriptOrigin::ScriptOrigin': Use constructor without the isolate. 
node_modules\nan\nan_scriptorigin.h(30,23): warning C4996: 'v8::ScriptOrigin::ScriptOrigin': Use constructor without the isolate. 

```

In Electron v32  it now throws these errors:

```
node_modules\nan\nan.h(700,39): error C2039: 'IdleNotificationDeadline': is not a member of 'v8::Isolate' 
node_modules\nan\nan.h(2560,8): error C2039: 'SetAccessor': is not a member of 'v8::ObjectTemplate' 
node_modules\nan\nan.h(2608,8): error C2039: 'SetAccessor': is not a member of 'v8::ObjectTemplate'
node_modules\nan\nan.h(2654,15): error C2039: 'SetAccessor': is not a member of 'v8::Object' 
node_modules\nan\nan.h(2730,56): error C2440: '<function-style-cast>': cannot convert
from 'initializer list' to 'v8::NamedPropertyHandlerConfiguration'
node_modules\nan\nan.h(2800,58): error C2440: '<function-style-cast>': cannot convert
from 'initializer list' to 'v8::IndexedPropertyHandlerConfiguration'

node_modules\nan\nan_scriptorigin.h(19,23): error C2664: 'v8::ScriptOrigin::ScriptOrig in(v8::Local<v8::Value>,int,int,bool,int,v8::Local<v8::Value>,bool,bool,bool,v8::Local<v8::Data>)': cannot convert argument 1 from 'v8::Isolate *' to 'v8::Local<v8::Value>' 
node_modules\nan\nan_scriptorigin.h(23,23): error C2664: 'v8::ScriptOrigin::ScriptOrig in(v8::Local<v8::Value>,int,int,bool,int,v8::Local<v8::Value>,bool,bool,bool,v8::Local<v8::Data>)': cannot convert argument 1 from 'v8::Isolate *' to 'v8::Local<v8::Value>' 
node_modules\nan\nan_scriptorigin.h(30,23): error C2664: 'v8::ScriptOrigin::ScriptOrig in(v8::Local<v8::Value>,int,int,bool,int,v8::Local<v8::Value>,bool,bool,bool,v8::Local<v8::Data>)': cannot convert argument 1 from 'v8::Isolate *' to 'v8::Local<v8::Value>' 


```

Could this used be to replace `IdleNotificationDeadline` in a different part of the code?

```
node_modules\nan\nan.h(700,39): warning C4996: 'v8::Isolate::IdleNotificationDeadline' : Use MemoryPressureNotification() to influence the GC schedule. 
```

Based on some comments in linked commits, MemoryPressurrNotification seems to correspond more to some GCLimit-ish-named thing not exposed by NAN. I do not see a straight replacement for this, so it will just be deprecated deplaced by a no-op.

On August 22, 2024 10:42:43 AM GMT+03:00, agracio ***@***.***> wrote:
>Could this used be to replace `IdleNotificationDeadline` in a different part of the code?
>
>```
>node_modules\nan\nan.h(700,39): warning C4996: 'v8::Isolate::IdleNotificationDeadline' : Use MemoryPressureNotification() to influence the GC schedule. 
>```
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/nodejs/nan/issues/973#issuecomment-2303996057
>You are receiving this because you commented.
>
>Message ID: ***@***.***>

Ok this one is on me, instead of just building your branch of `nan` and finding all the errors encountered on build I should have changed the files in my projects node_modules/nan

```
node_modules\nan\nan.h(2673,15): error C2664: 'v8::Maybe<bool> v8::Object::SetNativeDa taProperty(v8::Local<v8::Context>,v8::Local<v8::Name>,v8::AccessorNameGetterCallback,v8::AccessorNameSetterCallback,v8::Lo cal<v8::Value>,v8::PropertyAttribute,v8::SideEffectType,v8::SideEffectType)': cannot convert argument 6 from 'v8::AccessCo ntrol' to 'v8::PropertyAttribute' 
node_modules\nan\nan.h(2752,56): error C2440: '<function-style-cast>': cannot convertfrom 'initializer list' to 'v8::NamedPropertyHandlerConfiguration'
node_modules\nan\nan.h(2822,58): error C2440: '<function-style-cast>': cannot convert from 'initializer list' to 'v8::IndexedPropertyHandlerConfiguration' 
```

After the fix now getting even more very strange errors 

```
nan_callbacks_12_inl.h(277,34): error C2039: 'Get': is not a member o f 'Nan::ReturnValue<T>' [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(277,34): error C2039:         with [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(277,34): error C2039:         [ [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(277,34): error C2039:             T=v8::Value [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(277,34): error C2039:         ] [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(322,34): error C2039: 'Get': is not a member of 'Nan::ReturnValue<T>' [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(322,34): error C2039:         with [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(322,34): error C2039:         [ [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(322,34): error C2039:             T=v8::Value [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(322,34): error C2039:         ] [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(387,34): error C2039: 'Get': is not a member of 'Nan::ReturnValue<T>' [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(387,34): error C2039:         with [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(387,34): error C2039:         [ [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(387,34): error C2039:             T=v8::Boolean [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(387,34): error C2039:         ] [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(431,34): error C2039: 'Get': is not a member of 'Nan::ReturnValue<T>' [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(431,34): error C2039:         with [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(431,34): error C2039:         [ [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(431,34): error C2039:             T=v8::Integer [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(431,34): error C2039:         ] [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(562,34): error C2039: 'Get': is not a member of 'Nan::ReturnValue<T>' [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(562,34): error C2039:         with [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(562,34): error C2039:         [ [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(562,34): error C2039:             T=v8::Value [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(562,34): error C2039:         ] [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(606,34): error C2039: 'Get': is not a member of 'Nan::ReturnValue<T>' [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(606,34): error C2039:         with [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(606,34): error C2039:         [ [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(606,34): error C2039:             T=v8::Value [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(606,34): error C2039:         ] [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(671,34): error C2039: 'Get': is not a member of 'Nan::ReturnValue<T>' [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(671,34): error C2039:         with [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(671,34): error C2039:         [ [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(671,34): error C2039:             T=v8::Boolea n [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(671,34): error C2039:         ] [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(713,34): error C2039: 'Get': is not a member of 'Nan::ReturnValue<T>' [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(713,34): error C2039:         with [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(713,34): error C2039:         [ [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(713,34): error C2039:             T=v8::Integer [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan_callbacks_12_inl.h(713,34): error C2039:         ] [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan.h(2717,30): error C2072: 'deleter_': initialization of a function  [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan.h(2759,33): error C2065: 'deleter_': undeclared identifier [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
nan.h(2828,58): error C2440: '<function-style-cast>': cannot convert from 'initializer list' to 'v8::IndexedPropertyHandlerConfiguration' [C:\path\electron-edge-js\build\edge_coreclr.vcxproj]
```

Well here is a new list of errors

```
nan_callbacks_12_inl.h(19,35): error C7568: argument list missing after assumed function template 'ReturnValue' 
nan_callbacks_12_inl.h(19,29): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int 
nan_callbacks_12_inl.h(19,35): error C2144: syntax error: 'unknown-type' should be preceded by ')' 
nan_callbacks_12_inl.h(19,35): error C2144: syntax error: 'unknown-type' should be preceded by ';' 
nan_callbacks_12_inl.h(19,53): error C2059: syntax error: ')' 
nan_callbacks_12_inl.h(19,54): error C2988: unrecognizable template declaration/definition 
nan_callbacks_12_inl.h(19,54): error C2143: syntax error: missing ';'  before '<end Parse>' 
nan_callbacks_12_inl.h(20,1): error C2988: unrecognizable template de claration/definition 
nan_callbacks_12_inl.h(20,1): error C2059: syntax error: '}' 
nan_callbacks_12_inl.h(20,1): error C2143: syntax error: missing ';' before '}' 
nan_callbacks_12_inl.h(122,10): error C2977: 'Nan::imp::ReturnValue':  too many template arguments 
nan_callbacks_12_inl.h(122,25): error C2955: 'Nan::imp::ReturnValue':  use of class template requires template argument list 
nan_callbacks_12_inl.h(122,25): error C2027: use of undefined type 'Nan::imp::ReturnValue' 
nan_callbacks_12_inl.h(168,10): error C2977: 'Nan::imp::ReturnValue':  too many template arguments 
nan_callbacks_12_inl.h(168,25): error C2955: 'Nan::imp::ReturnValue':  use of class template requires template argument list 
nan_callbacks_12_inl.h(168,25): error C2027: use of undefined type 'Nan::imp::ReturnValue' 
nan_callbacks_12_inl.h(285,42): error C2977: 'Nan::imp::ReturnValue':  too many template arguments 
nan_callbacks_12_inl.h(285,58): error C2955: 'Nan::imp::ReturnValue':  use of class template requires template argument list 
nan_callbacks_12_inl.h(286,13): error C2027: use of undefined type 'Nan::imp::ReturnValue' 
nan_callbacks_12_inl.h(300,39): error C2664: 'void (v8::Local<v8::Str ing>,const Nan::PropertyCallbackInfo<v8::Value> &)': cannot convert argument 2 from 'Nan::imp::PropertyCallbackInfo<v8::Va lue>' to 'const Nan::PropertyCallbackInfo<v8::Value> &'  nan_callbacks_12_inl.h(301,37): error C2027: use of undefined type 'N an::imp::ReturnValue' 
nan_callbacks_12_inl.h(301,15): error C2672: 'Nan::imp::imp::GetReturnValue': no matching overloaded function found 
nan_callbacks_12_inl.h(345,46): error C2664: 'void (v8::Local<v8::String>,v8::Local<v8::Value>,const Nan::PropertyCallbackInfo<v8::Value> &)': cannot convert argument 3 from 'Nan::imp::Proper tyCallbackInfo<v8::Value>' to 'const Nan::PropertyCallbackInfo<v8::Value> &' 
nan_callbacks_12_inl.h(346,37): error C2027: use of undefined type 'Nan::imp::ReturnValue' 
nan_callbacks_12_inl.h(346,15): error C2672: 'Nan::imp::imp::GetReturnValue': no matching overloaded function found 
nan_callbacks_12_inl.h(388,12): error C2664: 'void (const Nan::PropertyCallbackInfo<v8::Array> &)': cannot convert argument 1 from 'Nan::imp::PropertyCallbackInfo<v8::Array>' to 'const Nan::P ropertyCallbackInfo<v8::Array> &' 
nan_callbacks_12_inl.h(410,39): error C2664: 'void (v8::Local<v8::String>,const Nan::PropertyCallbackInfo<v8::Boolean> &)': cannot convert argument 2 from 'Nan::imp::PropertyCallbackInfo<v8:: Boolean>' to 'const Nan::PropertyCallbackInfo<v8::Boolean> &' 
nan_callbacks_12_inl.h(411,37): error C2027: use of undefined type 'Nan::imp::ReturnValue' 
nan_callbacks_12_inl.h(411,15): error C2672: 'Nan::imp::imp::GetReturnValue': no matching overloaded function found 
nan_callbacks_12_inl.h(454,39): error C2664: 'void (v8::Local<v8::String>,const Nan::PropertyCallbackInfo<v8::Integer> &)': cannot convert argument 2 from 'Nan::imp::PropertyCallbackInfo<v8:: Integer>' to 'const Nan::PropertyCallbackInfo<v8::Integer> &' 
nan_callbacks_12_inl.h(455,37): error C2027: use of undefined type 'Nan::imp::ReturnValue' 
nan_callbacks_12_inl.h(455,15): error C2672: 'Nan::imp::imp::GetReturnValue': no matching overloaded function found 
nan_callbacks_12_inl.h(585,19): error C2664: 'void (uint32_t,const Nan::PropertyCallbackInfo<v8::Value> &)': cannot convert argument 2 from 'Nan::imp::PropertyCallbackInfo<v8::Value>' to 'con st Nan::PropertyCallbackInfo<v8::Value> &' 
nan_callbacks_12_inl.h(586,37): error C2027: use of undefined type 'Nan::imp::ReturnValue' 
nan_callbacks_12_inl.h(586,15): error C2672: 'Nan::imp::imp::GetReturnValue': no matching overloaded function found 
nan_callbacks_12_inl.h(629,26): error C2664: 'void (uint32_t,v8::Local<v8::Value>,const Nan::PropertyCallbackInfo<v8::Value> &)': cannot convert argument 3 from 'Nan::imp::PropertyCallbackInf o<v8::Value>' to 'const Nan::PropertyCallbackInfo<v8::Value> &' 
nan_callbacks_12_inl.h(630,37): error C2027: use of undefined type 'Nan::imp::ReturnValue' 
nan_callbacks_12_inl.h(630,15): error C2672: 'Nan::imp::imp::GetReturnValue': no matching overloaded function found 
nan_callbacks_12_inl.h(673,12): error C2664: 'void (const Nan::PropertyCallbackInfo<v8::Array> &)': cannot convert argument 1 from 'Nan::imp::PropertyCallbackInfo<v8::Array>' to 'const Nan::P ropertyCallbackInfo<v8::Array> &' 
nan_callbacks_12_inl.h(694,19): error C2664: 'void (uint32_t,const Nan::PropertyCallbackInfo<v8::Boolean> &)': cannot convert argument 2 from 'Nan::imp::PropertyCallbackInfo<v8::Boolean>' to 'const Nan::PropertyCallbackInfo<v8::Boolean> &' 
nan_callbacks_12_inl.h(695,37): error C2027: use of undefined type 'Nan::imp::ReturnValue' 
nan_callbacks_12_inl.h(695,15): error C2672: 'Nan::imp::imp::GetReturnValue': no matching overloaded function found 
nan_callbacks_12_inl.h(736,19): error C2664: 'void (uint32_t,const Nan::PropertyCallbackInfo<v8::Integer> &)': cannot convert argument 2 from 'Nan::imp::PropertyCallbackInfo<v8::Integer>' to 'const Nan::PropertyCallbackInfo<v8::Integer> &' 
nan_callbacks_12_inl.h(737,37): error C2027: use of undefined type 'Nan::imp::ReturnValue' 
nan_callbacks_12_inl.h(737,15): error C2672: 'Nan::imp::imp::GetReturnValue': no matching overloaded function found 
nan_new.h(21,15): error C2039: 'To': is not a member of 'Nan' 
nan_new.h(21,22): error C2275: 'v8::Integer': expected an expression instead of a type 
nan_new.h(21,34): error C2039: 'ToLocalChecked': is not a member of ' v8::Local<v8::Integer>' 
nan_new.h(28,15): error C2039: 'To': is not a member of 'Nan' 
nan_new.h(28,22): error C2275: 'v8::Int32': expected an expression instead of a type 
nan_new.h(28,32): error C2039: 'ToLocalChecked': is not a member of ' v8::Local<v8::Integer>' 
nan_new.h(35,15): error C2039: 'To': is not a member of 'Nan' 
nan_new.h(35,22): error C2275: 'v8::Uint32': expected an expression instead of a type 
nan_new.h(35,22): error C1003: error count exceeds 100; stopping compilation 
```

Getting closer

```
nan_callbacks_12_inl.h(344,13): error C2664: 'Nan::PropertyCallbackIn fo<v8::Value>::PropertyCallbackInfo(const v8::PropertyCallbackInfo<v8::Value> &,const v8::Local<v8::Value>)': cannot convert argument 1 from 'const v8::PropertyCallbackInfo<void>' to 'const v8::PropertyCallbackInfo<v8::Value> &' 
nan_callbacks_12_inl.h(628,13): error C2664: 'Nan::PropertyCallbackIn fo<v8::Value>::PropertyCallbackInfo(const v8::PropertyCallbackInfo<v8::Value> &,const v8::Local<v8::Value>)': cannot convert argument 1 from 'const v8::PropertyCallbackInfo<void>' to 'const v8::PropertyCallbackInfo<v8::Value> &' 
```

First of all wanted to say that I do appreciate the work, very quick responses and attempts to fix the error from everyone, in my previous posts I was rather frustrated that the problem is being ignored since `nan` only covers Node.js and other projects are left to fend for themselves.

Now that v8 12.8 was merged into main Node.js branch 23.0.0 I tested on a different project that does not depend on Electron and unfortunately getting exact same errors. 

Initially getting the same errors as with Electron build and after applying patch from `new_breakage` branch it is back to 
```
nan_callbacks_12_inl.h(344,13): error C2664: 'Nan::PropertyCallbackInfo<v8::Value>::PropertyCallbackInfo(const v8::PropertyCallbackInfo<v8::Value> &,const v8::Local<v8::Value>)': cannot convert argument 1 from 'const v8::PropertyCallbackInfo<void>' to 'const v8::PropertyCallbackInfo<v8::Value> &' 
nan_callbacks_12_inl.h(628,13): error C2664: 'Nan::PropertyCallbackInfo<v8::Value>::PropertyCallbackInfo(const v8::PropertyCallbackInfo<v8::Value> &,const v8::Local<v8::Value>)': cannot convert argument 1 from 'const v8::PropertyCallbackInfo<void>' to 'const v8::PropertyCallbackInfo<v8::Value> &' 
```

> And this issue is likely not only limited to Electron but to other apps that will slowly migrate to newer versions of v8.
> ... and when that happens I'm sure we will take another stab at it.😁

Node.js 23 release is a couple months away so while it does not require urgent attention perhaps there should be some plans to address it.

--- a/README.md
+++ b/README.md
@@ -297,7 +297,7 @@ The hooks to access V8 internals—including GC and statistics—are different a
  - <a href="doc/v8_internals.md#api_nan_set_counter_function"><b><code>Nan::SetCounterFunction()</code></b></a>
  - <a href="doc/v8_internals.md#api_nan_set_create_histogram_function"><b><code>Nan::SetCreateHistogramFunction()</code></b></a>
  - <a href="doc/v8_internals.md#api_nan_set_add_histogram_sample_function"><b><code>Nan::SetAddHistogramSampleFunction()</code></b></a>
- - <a href="doc/v8_internals.md#api_nan_idle_notification"><b><code>Nan::IdleNotification()</code></b></a>
+ - <a href="doc/v8_internals.md#api_nan_idle_notification"><del><b><code>Nan::IdleNotification()</code></b></del></a>
  - <a href="doc/v8_internals.md#api_nan_low_memory_notification"><b><code>Nan::LowMemoryNotification()</code></b></a>
  - <a href="doc/v8_internals.md#api_nan_context_disposed_notification"><b><code>Nan::ContextDisposedNotification()</code></b></a>
  - <a href="doc/v8_internals.md#api_nan_get_internal_field_pointer"><b><code>Nan::GetInternalFieldPointer()</code></b></a>

--- a/doc/v8_internals.md
+++ b/doc/v8_internals.md
@@ -11,7 +11,7 @@ The hooks to access V8 internals—including GC and statistics—are different a
  - <a href="#api_nan_set_counter_function"><b><code>Nan::SetCounterFunction()</code></b></a>
  - <a href="#api_nan_set_create_histogram_function"><b><code>Nan::SetCreateHistogramFunction()</code></b></a>
  - <a href="#api_nan_set_add_histogram_sample_function"><b><code>Nan::SetAddHistogramSampleFunction()</code></b></a>
- - <a href="#api_nan_idle_notification"><b><code>Nan::IdleNotification()</code></b></a>
+ - <a href="#api_nan_idle_notification"><del><b><code>Nan::IdleNotification()</code></b></del></a>
  - <a href="#api_nan_low_memory_notification"><b><code>Nan::LowMemoryNotification()</code></b></a>
  - <a href="#api_nan_context_disposed_notification"><b><code>Nan::ContextDisposedNotification()</code></b></a>
  - <a href="#api_nan_get_internal_field_pointer"><b><code>Nan::GetInternalFieldPointer()</code></b></a>
@@ -128,15 +128,15 @@ void Nan::SetAddHistogramSampleFunction(v8::AddHistogramSampleCallback cb)
 Calls V8's [`SetAddHistogramSampleFunction()`](https://v8docs.nodesource.com/node-8.16/d5/dda/classv8_1_1_isolate.html#aeb420b690bc2c216882d6fdd00ddd3ea).
 
 <a name="api_nan_idle_notification"></a>
-### Nan::IdleNotification()
+### <del>Nan::IdleNotification()</del>
 
 Signature:
 
 ```c++
-bool Nan::IdleNotification(int idle_time_in_ms)
+NAN_DEPRECATED bool Nan::IdleNotification(int idle_time_in_ms)
 ```
 
-Calls V8's [`IdleNotification()` or `IdleNotificationDeadline()`](https://v8docs.nodesource.com/node-8.16/d5/dda/classv8_1_1_isolate.html#ad6a2a02657f5425ad460060652a5a118) depending on V8 version.
+Calls V8's [`IdleNotification()` or `IdleNotificationDeadline()`](https://v8docs.nodesource.com/node-8.16/d5/dda/classv8_1_1_isolate.html#ad6a2a02657f5425ad460060652a5a118) depending on V8 version. Removed in V8 12.7.41.
 
 <a name="api_nan_low_memory_notification"></a>
 ### Nan::LowMemoryNotification()

--- a/nan.h
+++ b/nan.h
@@ -693,7 +693,8 @@ inline uv_loop_t* GetCurrentEventLoop() {
   void SetAddHistogramSampleFunction(v8::AddHistogramSampleCallback cb) {
     v8::Isolate::GetCurrent()->SetAddHistogramSampleFunction(cb);
   }
-
+#if defined(V8_MAJOR_VERSION) && \
+    ((V8_MAJOR_VERSION) <= 6 && V8_MINOR_VERSION < 5)
 #if defined(V8_MAJOR_VERSION) && (V8_MAJOR_VERSION > 4 ||                      \
   (V8_MAJOR_VERSION == 4 && defined(V8_MINOR_VERSION) && V8_MINOR_VERSION >= 3))
   inline bool IdleNotification(int idle_time_in_ms) {
@@ -704,6 +705,7 @@ inline uv_loop_t* GetCurrentEventLoop() {
   inline bool IdleNotification(int idle_time_in_ms) {
     return v8::Isolate::GetCurrent()->IdleNotification(idle_time_in_ms);
   }
+#endif
 #endif
 
   inline void LowMemoryNotification() {
@@ -1538,6 +1540,7 @@ typedef void (*FreeCallback)(char *data, void *hint);
 typedef const FunctionCallbackInfo<v8::Value>& NAN_METHOD_ARGS_TYPE;
 typedef void NAN_METHOD_RETURN_TYPE;
 
+
 typedef const PropertyCallbackInfo<v8::Value>& NAN_GETTER_ARGS_TYPE;
 typedef void NAN_GETTER_RETURN_TYPE;
 
@@ -1546,11 +1549,23 @@ typedef void NAN_SETTER_RETURN_TYPE;
 
 typedef const PropertyCallbackInfo<v8::Value>&
     NAN_PROPERTY_GETTER_ARGS_TYPE;
+#if defined(V8_MAJOR_VERSION) && (V8_MAJOR_VERSION > 12 ||                     \
+  (V8_MAJOR_VERSION == 12 && defined(V8_MINOR_VERSION) && V8_MINOR_VERSION > 4))
+typedef v8::Intercepted NAN_PROPERTY_GETTER_RETURN_TYPE;
+#else
 typedef void NAN_PROPERTY_GETTER_RETURN_TYPE;
+#endif
 
+#if defined(V8_MAJOR_VERSION) && (V8_MAJOR_VERSION > 12 ||                     \
+  (V8_MAJOR_VERSION == 12 && defined(V8_MINOR_VERSION) && V8_MINOR_VERSION > 4))
+typedef const PropertyCallbackInfo<void>&
+    NAN_PROPERTY_SETTER_ARGS_TYPE;
+typedef v8::Intercepted NAN_PROPERTY_SETTER_RETURN_TYPE;
+#else
 typedef const PropertyCallbackInfo<v8::Value>&
     NAN_PROPERTY_SETTER_ARGS_TYPE;
 typedef void NAN_PROPERTY_SETTER_RETURN_TYPE;
+#endif
 
 typedef const PropertyCallbackInfo<v8::Array>&
     NAN_PROPERTY_ENUMERATOR_ARGS_TYPE;
@@ -1558,29 +1573,68 @@ typedef void NAN_PROPERTY_ENUMERATOR_RETURN_TYPE;
 
 typedef const PropertyCallbackInfo<v8::Boolean>&
     NAN_PROPERTY_DELETER_ARGS_TYPE;
+
+#if defined(V8_MAJOR_VERSION) && (V8_MAJOR_VERSION > 12 ||                     \
+  (V8_MAJOR_VERSION == 12 && defined(V8_MINOR_VERSION) && V8_MINOR_VERSION > 4))
+typedef v8::Intercepted NAN_PROPERTY_DELETER_RETURN_TYPE;
+#else
 typedef void NAN_PROPERTY_DELETER_RETURN_TYPE;
+#endif
+
 
 typedef const PropertyCallbackInfo<v8::Integer>&
     NAN_PROPERTY_QUERY_ARGS_TYPE;
+#if defined(V8_MAJOR_VERSION) && (V8_MAJOR_VERSION > 12 ||                     \
+  (V8_MAJOR_VERSION == 12 && defined(V8_MINOR_VERSION) && V8_MINOR_VERSION > 4))
+typedef v8::Intercepted NAN_PROPERTY_QUERY_RETURN_TYPE;
+#else
 typedef void NAN_PROPERTY_QUERY_RETURN_TYPE;
+#endif
 
 typedef const PropertyCallbackInfo<v8::Value>& NAN_INDEX_GETTER_ARGS_TYPE;
+#if defined(V8_MAJOR_VERSION) && (V8_MAJOR_VERSION > 12 ||                     \
+  (V8_MAJOR_VERSION == 12 && defined(V8_MINOR_VERSION) && V8_MINOR_VERSION > 4))
+typedef v8::Intercepted NAN_INDEX_GETTER_RETURN_TYPE;
+#else
 typedef void NAN_INDEX_GETTER_RETURN_TYPE;
+#endif
 
+
+#if defined(V8_MAJOR_VERSION) && (V8_MAJOR_VERSION > 12 ||                     \
+  (V8_MAJOR_VERSION == 12 && defined(V8_MINOR_VERSION) && V8_MINOR_VERSION > 4))
+typedef const PropertyCallbackInfo<void>& NAN_INDEX_SETTER_ARGS_TYPE;
+typedef v8::Intercepted NAN_INDEX_SETTER_RETURN_TYPE;
+#else
 typedef const PropertyCallbackInfo<v8::Value>& NAN_INDEX_SETTER_ARGS_TYPE;
 typedef void NAN_INDEX_SETTER_RETURN_TYPE;
+#endif
 
 typedef const PropertyCallbackInfo<v8::Array>&
     NAN_INDEX_ENUMERATOR_ARGS_TYPE;
+#if defined(V8_MAJOR_VERSION) && (V8_MAJOR_VERSION > 12 ||                     \
+  (V8_MAJOR_VERSION == 12 && defined(V8_MINOR_VERSION) && V8_MINOR_VERSION > 4))
+typedef v8::Intercepted NAN_INDEX_ENUMERATOR_RETURN_TYPE;
+#else
 typedef void NAN_INDEX_ENUMERATOR_RETURN_TYPE;
+#endif
 
 typedef const PropertyCallbackInfo<v8::Boolean>&
     NAN_INDEX_DELETER_ARGS_TYPE;
+#if defined(V8_MAJOR_VERSION) && (V8_MAJOR_VERSION > 12 ||                     \
+  (V8_MAJOR_VERSION == 12 && defined(V8_MINOR_VERSION) && V8_MINOR_VERSION > 4))
+typedef v8::Intercepted NAN_INDEX_DELETER_RETURN_TYPE;
+#else
 typedef void NAN_INDEX_DELETER_RETURN_TYPE;
+#endif
 
 typedef const PropertyCallbackInfo<v8::Integer>&
     NAN_INDEX_QUERY_ARGS_TYPE;
+#if defined(V8_MAJOR_VERSION) && (V8_MAJOR_VERSION > 12 ||                     \
+  (V8_MAJOR_VERSION == 12 && defined(V8_MINOR_VERSION) && V8_MINOR_VERSION > 4))
+typedef v8::Intercepted NAN_INDEX_QUERY_RETURN_TYPE;
+#else
 typedef void NAN_INDEX_QUERY_RETURN_TYPE;
+#endif
 
 #define NAN_METHOD(name)                                                       \
     Nan::NAN_METHOD_RETURN_TYPE name(Nan::NAN_METHOD_ARGS_TYPE info)
@@ -2522,7 +2576,8 @@ inline void SetPrototypeMethod(
 }
 
 //=== Accessors and Such =======================================================
-
+#if defined(V8_MAJOR_VERSION) && \
+    ((V8_MAJOR_VERSION) <= 6 && V8_MINOR_VERSION < 5)
 NAN_DEPRECATED inline void SetAccessor(
     v8::Local<v8::ObjectTemplate> tpl
   , v8::Local<v8::String> name
@@ -2669,6 +2724,7 @@ inline bool SetAccessor(
     , attribute);
 #endif
 }
+#endif
 
 inline void SetNamedPropertyHandler(
     v8::Local<v8::ObjectTemplate> tpl

--- a/nan.h
+++ b/nan.h
@@ -2576,8 +2576,7 @@ inline void SetPrototypeMethod(
 }
 
 //=== Accessors and Such =======================================================
-#if defined(V8_MAJOR_VERSION) && \
-    ((V8_MAJOR_VERSION) <= 6 && V8_MINOR_VERSION < 5)
+
 NAN_DEPRECATED inline void SetAccessor(
     v8::Local<v8::ObjectTemplate> tpl
   , v8::Local<v8::String> name
@@ -2724,7 +2723,6 @@ inline bool SetAccessor(
     , attribute);
 #endif
 }
-#endif
 
 inline void SetNamedPropertyHandler(
     v8::Local<v8::ObjectTemplate> tpl

--- a/nan.h
+++ b/nan.h
@@ -693,19 +693,24 @@ inline uv_loop_t* GetCurrentEventLoop() {
   void SetAddHistogramSampleFunction(v8::AddHistogramSampleCallback cb) {
     v8::Isolate::GetCurrent()->SetAddHistogramSampleFunction(cb);
   }
-#if defined(V8_MAJOR_VERSION) && \
-    ((V8_MAJOR_VERSION) <= 6 && V8_MINOR_VERSION < 5)
-#if defined(V8_MAJOR_VERSION) && (V8_MAJOR_VERSION > 4 ||                      \
+
+#if defined(V8_MAJOR_VERSION) &&                                               \
+      (V8_MAJOR_VERSION > 12 ||                                                \
+       (V8_MAJOR_VERSION == 12 && defined(V8_MINOR_VERSION) &&                 \
+        V8_MINOR_VERSION >= 7))
+  NAN_DEPRECATED inline bool IdleNotification(int) {
+    return true;
+  }
+# elif defined(V8_MAJOR_VERSION) && (V8_MAJOR_VERSION > 4 ||                   \
   (V8_MAJOR_VERSION == 4 && defined(V8_MINOR_VERSION) && V8_MINOR_VERSION >= 3))
-  inline bool IdleNotification(int idle_time_in_ms) {
+  NAN_DEPRECATED inline bool IdleNotification(int idle_time_in_ms) {
     return v8::Isolate::GetCurrent()->IdleNotificationDeadline(
         idle_time_in_ms * 0.001);
   }
 # else
-  inline bool IdleNotification(int idle_time_in_ms) {
+  NAN_DEPRECATED inline bool IdleNotification(int idle_time_in_ms) {
     return v8::Isolate::GetCurrent()->IdleNotification(idle_time_in_ms);
   }
-#endif
 #endif
 
   inline void LowMemoryNotification() {
@@ -1539,7 +1544,6 @@ typedef void (*FreeCallback)(char *data, void *hint);
 
 typedef const FunctionCallbackInfo<v8::Value>& NAN_METHOD_ARGS_TYPE;
 typedef void NAN_METHOD_RETURN_TYPE;
-
 
 typedef const PropertyCallbackInfo<v8::Value>& NAN_GETTER_ARGS_TYPE;
 typedef void NAN_GETTER_RETURN_TYPE;

--- a/nan_callbacks.h
+++ b/nan_callbacks.h
@@ -20,6 +20,17 @@ typedef void(*SetterCallback)(
     v8::Local<v8::String>,
     v8::Local<v8::Value>,
     const PropertyCallbackInfo<void>&);
+
+#if defined(V8_MAJOR_VERSION) && (V8_MAJOR_VERSION > 12 ||                     \
+  (V8_MAJOR_VERSION == 12 && defined(V8_MINOR_VERSION) && V8_MINOR_VERSION > 4))
+typedef v8::Intercepted(*PropertyGetterCallback)(
+    v8::Local<v8::String>,
+    const PropertyCallbackInfo<v8::Value>&);
+typedef v8::Intercepted(*PropertySetterCallback)(
+    v8::Local<v8::String>,
+    v8::Local<v8::Value>,
+    const PropertyCallbackInfo<void>&);
+#else
 typedef void(*PropertyGetterCallback)(
     v8::Local<v8::String>,
     const PropertyCallbackInfo<v8::Value>&);
@@ -27,8 +38,33 @@ typedef void(*PropertySetterCallback)(
     v8::Local<v8::String>,
     v8::Local<v8::Value>,
     const PropertyCallbackInfo<v8::Value>&);
+#endif
 typedef void(*PropertyEnumeratorCallback)
     (const PropertyCallbackInfo<v8::Array>&);
+#if defined(V8_MAJOR_VERSION) && (V8_MAJOR_VERSION > 12 ||                     \
+  (V8_MAJOR_VERSION == 12 && defined(V8_MINOR_VERSION) && V8_MINOR_VERSION > 4))
+typedef v8::Intercepted(*PropertyDeleterCallback)(
+    v8::Local<v8::String>,
+    const PropertyCallbackInfo<v8::Boolean>&);
+typedef v8::Intercepted(*PropertyQueryCallback)(
+    v8::Local<v8::String>,
+    const PropertyCallbackInfo<v8::Integer>&);
+typedef v8::Intercepted(*IndexGetterCallback)(
+    uint32_t,
+    const PropertyCallbackInfo<v8::Value>&);
+typedef v8::Intercepted(*IndexSetterCallback)(
+    uint32_t,
+    v8::Local<v8::Value>,
+    const PropertyCallbackInfo<void>&);
+typedef v8::Intercepted(*IndexEnumeratorCallback)
+    (const PropertyCallbackInfo<v8::Array>&);
+typedef v8::Intercepted(*IndexDeleterCallback)(
+    uint32_t,
+    const PropertyCallbackInfo<v8::Boolean>&);
+typedef v8::Intercepted(*IndexQueryCallback)(
+    uint32_t,
+    const PropertyCallbackInfo<v8::Integer>&);
+#else
 typedef void(*PropertyDeleterCallback)(
     v8::Local<v8::String>,
     const PropertyCallbackInfo<v8::Boolean>&);
@@ -50,7 +86,7 @@ typedef void(*IndexDeleterCallback)(
 typedef void(*IndexQueryCallback)(
     uint32_t,
     const PropertyCallbackInfo<v8::Integer>&);
-
+#endif
 namespace imp {
 #if (NODE_MODULE_VERSION < NODE_16_0_MODULE_VERSION)
 typedef v8::Local<v8::AccessorSignature> Sig;

--- a/nan_callbacks_12_inl.h
+++ b/nan_callbacks_12_inl.h
@@ -297,7 +297,7 @@ typedef v8::Intercepted (*NativePropertySetter)(
   , const v8::PropertyCallbackInfo<void> &);
 
 #else
-
+static
 void PropertyGetterCallbackWrapper(
     v8::Local<v8::Name> property
   , const v8::PropertyCallbackInfo<v8::Value> &info) {

--- a/nan_callbacks_12_inl.h
+++ b/nan_callbacks_12_inl.h
@@ -671,7 +671,6 @@ typedef void (*NativeIndexQuery)
     (uint32_t, const v8::PropertyCallbackInfo<v8::Integer> &);
 
 #endif
-}
-// end of namespace imp
+}  // end of namespace imp
 
 #endif  // NAN_CALLBACKS_12_INL_H_

--- a/nan_callbacks_12_inl.h
+++ b/nan_callbacks_12_inl.h
@@ -256,7 +256,48 @@ typedef void (*NativeSetter)(
 #endif
 
 #if NODE_MODULE_VERSION > NODE_0_12_MODULE_VERSION
+
+#if defined(V8_MAJOR_VERSION) && (V8_MAJOR_VERSION > 12 ||                     \
+  (V8_MAJOR_VERSION == 12 && defined(V8_MINOR_VERSION) && V8_MINOR_VERSION > 4))
 static
+v8::Intercepted PropertyGetterCallbackWrapper(
+    v8::Local<v8::Name> property
+  , const v8::PropertyCallbackInfo<v8::Value> &info) {
+  v8::Local<v8::Object> obj = info.Data().As<v8::Object>();
+  PropertyCallbackInfo<v8::Value>
+      cbinfo(info, obj->GetInternalField(kDataIndex).As<v8::Value>());
+  PropertyGetterCallback callback = reinterpret_cast<PropertyGetterCallback>(
+      reinterpret_cast<intptr_t>(
+          obj->GetInternalField(kPropertyGetterIndex)
+              .As<v8::Value>().As<v8::External>()->Value()));
+  return callback(property.As<v8::String>(), cbinfo);
+}
+
+typedef v8::Intercepted (*NativePropertyGetter)
+    (v8::Local<v8::Name>, const v8::PropertyCallbackInfo<v8::Value> &);
+
+static
+v8::Intercepted PropertySetterCallbackWrapper(
+    v8::Local<v8::Name> property
+  , v8::Local<v8::Value> value
+  , const v8::PropertyCallbackInfo<void> &info) {
+  v8::Local<v8::Object> obj = info.Data().As<v8::Object>();
+  PropertyCallbackInfo<void>
+      cbinfo(info, obj->GetInternalField(kDataIndex).As<v8::Value>());
+  PropertySetterCallback callback = reinterpret_cast<PropertySetterCallback>(
+      reinterpret_cast<intptr_t>(
+          obj->GetInternalField(kPropertySetterIndex)
+              .As<v8::Value>().As<v8::External>()->Value()));
+  return callback(property.As<v8::String>(), value, cbinfo);
+}
+
+typedef v8::Intercepted (*NativePropertySetter)(
+    v8::Local<v8::Name>
+  , v8::Local<v8::Value>
+  , const v8::PropertyCallbackInfo<void> &);
+
+#else
+
 void PropertyGetterCallbackWrapper(
     v8::Local<v8::Name> property
   , const v8::PropertyCallbackInfo<v8::Value> &info) {
@@ -292,6 +333,7 @@ typedef void (*NativePropertySetter)(
     v8::Local<v8::Name>
   , v8::Local<v8::Value>
   , const v8::PropertyCallbackInfo<v8::Value> &);
+#endif
 
 static
 void PropertyEnumeratorCallbackWrapper(
@@ -309,6 +351,43 @@ void PropertyEnumeratorCallbackWrapper(
 typedef void (*NativePropertyEnumerator)
     (const v8::PropertyCallbackInfo<v8::Array> &);
 
+#if defined(V8_MAJOR_VERSION) && (V8_MAJOR_VERSION > 12 ||                     \
+  (V8_MAJOR_VERSION == 12 && defined(V8_MINOR_VERSION) && V8_MINOR_VERSION > 4))
+static
+v8::Intercepted PropertyDeleterCallbackWrapper(
+    v8::Local<v8::Name> property
+  , const v8::PropertyCallbackInfo<v8::Boolean> &info) {
+  v8::Local<v8::Object> obj = info.Data().As<v8::Object>();
+  PropertyCallbackInfo<v8::Boolean>
+      cbinfo(info, obj->GetInternalField(kDataIndex).As<v8::Value>());
+  PropertyDeleterCallback callback = reinterpret_cast<PropertyDeleterCallback>(
+      reinterpret_cast<intptr_t>(
+          obj->GetInternalField(kPropertyDeleterIndex)
+              .As<v8::Value>().As<v8::External>()->Value()));
+  return callback(property.As<v8::String>(), cbinfo);
+}
+
+typedef v8::Intercepted (NativePropertyDeleter)
+    (v8::Local<v8::Name>, const v8::PropertyCallbackInfo<v8::Boolean> &);
+
+
+static
+v8::Intercepted PropertyQueryCallbackWrapper(
+    v8::Local<v8::Name> property
+  , const v8::PropertyCallbackInfo<v8::Integer> &info) {
+  v8::Local<v8::Object> obj = info.Data().As<v8::Object>();
+  PropertyCallbackInfo<v8::Integer>
+      cbinfo(info, obj->GetInternalField(kDataIndex).As<v8::Value>());
+  PropertyQueryCallback callback = reinterpret_cast<PropertyQueryCallback>(
+      reinterpret_cast<intptr_t>(
+          obj->GetInternalField(kPropertyQueryIndex)
+              .As<v8::Value>().As<v8::External>()->Value()));
+  return callback(property.As<v8::String>(), cbinfo);
+}
+
+typedef v8::Intercepted (*NativePropertyQuery)
+    (v8::Local<v8::Name>, const v8::PropertyCallbackInfo<v8::Integer> &);
+#else
 static
 void PropertyDeleterCallbackWrapper(
     v8::Local<v8::Name> property
@@ -342,6 +421,7 @@ void PropertyQueryCallbackWrapper(
 
 typedef void (*NativePropertyQuery)
     (v8::Local<v8::Name>, const v8::PropertyCallbackInfo<v8::Integer> &);
+#endif
 #else
 static
 void PropertyGetterCallbackWrapper(
@@ -431,6 +511,45 @@ typedef void (*NativePropertyQuery)
     (v8::Local<v8::String>, const v8::PropertyCallbackInfo<v8::Integer> &);
 #endif
 
+#if defined(V8_MAJOR_VERSION) && (V8_MAJOR_VERSION > 12 ||                     \
+  (V8_MAJOR_VERSION == 12 && defined(V8_MINOR_VERSION) && V8_MINOR_VERSION > 4))
+static
+v8::Intercepted IndexGetterCallbackWrapper(
+    uint32_t index, const v8::PropertyCallbackInfo<v8::Value> &info) {
+  v8::Local<v8::Object> obj = info.Data().As<v8::Object>();
+  PropertyCallbackInfo<v8::Value>
+      cbinfo(info, obj->GetInternalField(kDataIndex).As<v8::Value>());
+  IndexGetterCallback callback = reinterpret_cast<IndexGetterCallback>(
+      reinterpret_cast<intptr_t>(
+          obj->GetInternalField(kIndexPropertyGetterIndex)
+              .As<v8::Value>().As<v8::External>()->Value()));
+  return callback(index, cbinfo);
+}
+
+typedef v8::Intercepted (*NativeIndexGetter)
+    (uint32_t, const v8::PropertyCallbackInfo<v8::Value> &);
+
+static
+v8::Intercepted IndexSetterCallbackWrapper(
+    uint32_t index
+  , v8::Local<v8::Value> value
+  , const v8::PropertyCallbackInfo<void> &info) {
+  v8::Local<v8::Object> obj = info.Data().As<v8::Object>();
+  PropertyCallbackInfo<void>
+      cbinfo(info, obj->GetInternalField(kDataIndex).As<v8::Value>());
+  IndexSetterCallback callback = reinterpret_cast<IndexSetterCallback>(
+      reinterpret_cast<intptr_t>(
+          obj->GetInternalField(kIndexPropertySetterIndex)
+              .As<v8::Value>().As<v8::External>()->Value()));
+  return callback(index, value, cbinfo);
+}
+
+typedef v8::Intercepted (*NativeIndexSetter)(
+    uint32_t
+  , v8::Local<v8::Value>
+  , const v8::PropertyCallbackInfo<void> &);
+
+#else
 static
 void IndexGetterCallbackWrapper(
     uint32_t index, const v8::PropertyCallbackInfo<v8::Value> &info) {
@@ -446,7 +565,6 @@ void IndexGetterCallbackWrapper(
 
 typedef void (*NativeIndexGetter)
     (uint32_t, const v8::PropertyCallbackInfo<v8::Value> &);
-
 static
 void IndexSetterCallbackWrapper(
     uint32_t index
@@ -467,6 +585,8 @@ typedef void (*NativeIndexSetter)(
   , v8::Local<v8::Value>
   , const v8::PropertyCallbackInfo<v8::Value> &);
 
+#endif
+
 static
 void IndexEnumeratorCallbackWrapper(
     const v8::PropertyCallbackInfo<v8::Array> &info) {
@@ -484,6 +604,40 @@ void IndexEnumeratorCallbackWrapper(
 typedef void (*NativeIndexEnumerator)
     (const v8::PropertyCallbackInfo<v8::Array> &);
 
+#if defined(V8_MAJOR_VERSION) && (V8_MAJOR_VERSION > 12 ||                     \
+  (V8_MAJOR_VERSION == 12 && defined(V8_MINOR_VERSION) && V8_MINOR_VERSION > 4))
+static
+v8::Intercepted IndexDeleterCallbackWrapper(
+    uint32_t index, const v8::PropertyCallbackInfo<v8::Boolean> &info) {
+  v8::Local<v8::Object> obj = info.Data().As<v8::Object>();
+  PropertyCallbackInfo<v8::Boolean>
+      cbinfo(info, obj->GetInternalField(kDataIndex).As<v8::Value>());
+  IndexDeleterCallback callback = reinterpret_cast<IndexDeleterCallback>(
+      reinterpret_cast<intptr_t>(
+          obj->GetInternalField(kIndexPropertyDeleterIndex)
+              .As<v8::Value>().As<v8::External>()->Value()));
+  return callback(index, cbinfo);
+}
+
+typedef v8::Intercepted (*NativeIndexDeleter)
+    (uint32_t, const v8::PropertyCallbackInfo<v8::Boolean> &);
+
+static
+v8::Intercepted IndexQueryCallbackWrapper(
+    uint32_t index, const v8::PropertyCallbackInfo<v8::Integer> &info) {
+  v8::Local<v8::Object> obj = info.Data().As<v8::Object>();
+  PropertyCallbackInfo<v8::Integer>
+      cbinfo(info, obj->GetInternalField(kDataIndex).As<v8::Value>());
+  IndexQueryCallback callback = reinterpret_cast<IndexQueryCallback>(
+      reinterpret_cast<intptr_t>(
+          obj->GetInternalField(kIndexPropertyQueryIndex)
+              .As<v8::Value>().As<v8::External>()->Value()));
+  return callback(index, cbinfo);
+}
+
+typedef v8::Intercepted (*NativeIndexQuery)
+    (uint32_t, const v8::PropertyCallbackInfo<v8::Integer> &);
+#else
 static
 void IndexDeleterCallbackWrapper(
     uint32_t index, const v8::PropertyCallbackInfo<v8::Boolean> &info) {
@@ -515,6 +669,9 @@ void IndexQueryCallbackWrapper(
 
 typedef void (*NativeIndexQuery)
     (uint32_t, const v8::PropertyCallbackInfo<v8::Integer> &);
-}  // end of namespace imp
+
+#endif
+}
+// end of namespace imp
 
 #endif  // NAN_CALLBACKS_12_INL_H_

--- a/nan_scriptorigin.h
+++ b/nan_scriptorigin.h
@@ -11,7 +11,24 @@
 
 class ScriptOrigin : public v8::ScriptOrigin {
  public:
-#if defined(V8_MAJOR_VERSION) && (V8_MAJOR_VERSION > 9 ||                      \
+
+#if defined(V8_MAJOR_VERSION) && (V8_MAJOR_VERSION > 11 \
+    && defined(V8_MINOR_VERSION) && V8_MINOR_VERSION > 7)
+  explicit ScriptOrigin(v8::Local<v8::Value> name) :
+      v8::ScriptOrigin(name) {}
+
+  ScriptOrigin(v8::Local<v8::Value> name
+             , v8::Local<v8::Integer> line) :
+      v8::ScriptOrigin(name
+                   , To<int32_t>(line).FromMaybe(0)) {}
+
+  ScriptOrigin(v8::Local<v8::Value> name
+             , v8::Local<v8::Integer> line
+             , v8::Local<v8::Integer> column) :
+      v8::ScriptOrigin(name
+                   , To<int32_t>(line).FromMaybe(0)
+                   , To<int32_t>(column).FromMaybe(0)) {}
+#elif defined(V8_MAJOR_VERSION) && (V8_MAJOR_VERSION > 9 ||                      \
   (V8_MAJOR_VERSION == 9 && (defined(V8_MINOR_VERSION) && (V8_MINOR_VERSION > 0\
       || (V8_MINOR_VERSION == 0 && defined(V8_BUILD_NUMBER)                    \
           && V8_BUILD_NUMBER >= 1)))))

